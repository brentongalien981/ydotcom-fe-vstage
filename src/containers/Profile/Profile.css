.profile-container {
  text-align: center;
  background-color: #fff;
  margin-bottom: 30px;
  padding: 30px;
  border-radius: 10px;
  box-shadow: 0px 0px 20px rgb(210, 210, 210);
}

.profile-container .page-title {
  text-align: left;
}

.profile-sections {
  margin: 10px 0;
  /* background-color: #FFF; */
}

.profile-container .profile-photo {
  width: 100px;
  border: 1px solid #DDD;
}