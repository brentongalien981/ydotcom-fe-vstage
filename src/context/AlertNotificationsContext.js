import React, { useState, createContext, useContext, useEffect } from 'react';
import MyJsonLocalStorage from '../utils/MyJsonLocalStorage';



const AlertNotificationsContext = createContext();


export const AlertNotificationsProvider = ({ children }) => {

  const [alertNotifications, setAlertNotifications] = useState([]);


  useEffect(() => {
    // Check existing alertNotifications from localStorage.
    const storedAlertNotifications = MyJsonLocalStorage.get("alertNotifications");

    if (storedAlertNotifications?.length > 0) {
      setAlertNotifications(storedAlertNotifications);
    }

  }, []);


  const addAlertNotification = (heading, message, variant) => {
    const id = Date.now().toString();
    const updatedAlertNotifications = [...alertNotifications, { id, heading, message, variant }];
    setAlertNotifications(updatedAlertNotifications);
    MyJsonLocalStorage.set("alertNotifications", updatedAlertNotifications);
  };


  const addAlertNotificationByProps = ({ heading, message, variant }) => {
        
    const id = Date.now().toString();
    const updatedAlertNotifications = [...alertNotifications, { id, heading, message, variant }];
    setAlertNotifications(updatedAlertNotifications);
    MyJsonLocalStorage.set("alertNotifications", updatedAlertNotifications);
  };




  const removeAlertNotification = (id) => {
    const updatedAlertNotifications = alertNotifications.filter((notification) => notification.id !== id);
    setAlertNotifications([...updatedAlertNotifications]);
    MyJsonLocalStorage.set("alertNotifications", updatedAlertNotifications);
  };


  return (
    <AlertNotificationsContext.Provider value={{ alertNotifications, addAlertNotification, addAlertNotificationByProps, removeAlertNotification }}>
      {children}
    </AlertNotificationsContext.Provider>
  );

};


export const useAlertNotifications = () => useContext(AlertNotificationsContext);
