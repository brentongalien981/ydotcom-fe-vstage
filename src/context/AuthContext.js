import React, { createContext, useState, useContext, useEffect } from "react";
import MyJsonLocalStorage from "../utils/MyJsonLocalStorage";


const AuthContext = createContext();


export const AuthProvider = ({ children }) => {

  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [username, setUsername] = useState("");
  const [token, setToken] = useState(null);
  const [profilePhotoSource, setProfilePhotoSource] = useState(null);


  useEffect(() => {
    // Check existing auth credentials from localStorage.
    const storedAuth = MyJsonLocalStorage.get("auth");

    if (storedAuth) {
      setIsLoggedIn(storedAuth.isLoggedIn);
      setUsername(storedAuth.username);
      setToken(storedAuth.token);
      setProfilePhotoSource(storedAuth.profilePhotoSource);
    }

  }, []);


  const login = (username, token, profilePhotoSource) => {
    setUsername(username);
    setToken(token);
    setIsLoggedIn(true);
    setProfilePhotoSource(profilePhotoSource);

    // Store auth credentials in localStorage.
    MyJsonLocalStorage.set("auth", { username, isLoggedIn: true, token, profilePhotoSource });
  };


  const logout = () => {
    setUsername("");
    setToken(null);
    setIsLoggedIn(false);
    setProfilePhotoSource(null);

    // Reset auth credentials in localStorage.
    MyJsonLocalStorage.set("auth", null);
  };


  return (
    <AuthContext.Provider value={{ isLoggedIn, username, token, profilePhotoSource, login, logout }}>
      {children}
    </AuthContext.Provider>
  );

};


export const useAuth = () => {
  return useContext(AuthContext);
};