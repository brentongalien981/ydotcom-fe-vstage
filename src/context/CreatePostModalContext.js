import React, { createContext, useContext, useState } from "react";


const CreatePostModalContext = createContext();


export const CreatePostModalProvider = ({ children }) => {

  const [isModalShown, setIsModalShown] = useState(false);


  const hideCreatePostModal = () => {
    setIsModalShown(false);
  };


  const showCreatePostModal = () => {
    setIsModalShown(true);
  };


  return (
    <CreatePostModalContext.Provider value={{ isModalShown, hideCreatePostModal, showCreatePostModal }}>
      {children}
    </CreatePostModalContext.Provider>
  );

};


export const useCreatePostModal = () => {
  return useContext(CreatePostModalContext);
};