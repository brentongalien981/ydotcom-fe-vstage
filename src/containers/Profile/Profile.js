import { Button, Image, Spinner, Tab, Tabs } from "react-bootstrap";
import "./Profile.css";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import { selectorIsAuthFollowingTheUser, selectorIsReadingProfile, selectorProfile, selectorReadProfileError } from "../../redux/selectors/profileSelectors";
import "animate.css";
import { followUser, readProfile } from "../../redux/actions/profileActions";
import Bio from "../../components/Bio/Bio";
import { useAlertNotifications } from "../../context/AlertNotificationsContext";


const Profile = () => {

  const dispatch = useDispatch();
  const { username } = useParams();
  const isReadingProfile = useSelector(selectorIsReadingProfile);
  const readError = useSelector(selectorReadProfileError);
  const profile = useSelector(selectorProfile);
  const isAuthFollowingTheUser = useSelector(selectorIsAuthFollowingTheUser)
  const { addAlertNotification } = useAlertNotifications();


  useEffect(() => {
    dispatch(readProfile(username));

  }, [dispatch, username]);


  function handleFollowUser() {
    dispatch(followUser(profile.userId, addAlertNotification))
  }


  
  function getFollowOrUnfollowBtn() {

    const eventHandler = isAuthFollowingTheUser ? () => { alert("LATER: This feature is pending...") } : handleFollowUser;
    const btnText = isAuthFollowingTheUser ? "Unfollow" : "Follow";
    const btnVariant = isAuthFollowingTheUser ? "warning" : "primary";

    let btn = (
      <div className="profile-sections">
        <Button className={`btn btn-${btnVariant}`} onClick={eventHandler}>{btnText}</Button>
      </div>
    );

    return btn;
  }



  // Set loader component.
  let loaderComponent = null;
  if (isReadingProfile) {
    loaderComponent = (
      <div className="d-flex justify-content-center" style={{ padding: "20px" }}>
        <Spinner animation="border" variant="primary" className="animate__animated animate__fadeIn" data-testid="theSpinner" />
      </div>
    );
  }


  // Set error component.
  let errorComponent = null;
  if (readError) {
    errorComponent = (<p>{readError}</p>);
  }


  // Set profile component.
  let profileComponent = null;

  if (!isReadingProfile && !readError) {

    const profilePhotoSrc = profile.photoSource ? `/photos/${profile.photoSource}` : `/photos/penguin2.jpg`;

    profileComponent = (
      <>

        {/* Profile Photo */}
        <div className="profile-sections">
          <Image src={profilePhotoSrc} roundedCircle className="profile-photo" />
        </div>

        {/* Username */}
        <div className="profile-sections">@{profile.username}</div>

        {/* Follow Button */}
        {getFollowOrUnfollowBtn()}

        {/* Profile Tabs */}
        <Tabs
          defaultActiveKey="bio"
          className="mb-3 profile-sections"
          fill
        >
          <Tab eventKey="bio" title="Bio"><Bio /></Tab>
          <Tab eventKey="posts" title="Posts">Tab content for Posts</Tab>
          <Tab eventKey="followers" title="Followers">Tab content for Followers</Tab>
          <Tab eventKey="following" title="Following">Tab content for Following</Tab>
        </Tabs>

      </>
    );
  }



  return (
    <div className="profile-container">
      <h4 className="page-title">Profile</h4>
      {loaderComponent}
      {errorComponent}
      {profileComponent}
    </div>
  );
};


export default Profile;