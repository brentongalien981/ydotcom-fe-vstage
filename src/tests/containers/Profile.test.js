import React from 'react'
import { act, fireEvent, screen } from '@testing-library/react'

// We're using our own custom render function and not RTL's render.
import { renderWithProviders } from '../../utils/test-utils/renderWithProviders'

import Profile from '../../containers/Profile/Profile';
import { BrowserRouter } from 'react-router-dom/cjs/react-router-dom.min';
import { AlertNotificationsProvider } from '../../context/AlertNotificationsContext';

afterEach(() => {
  jest.restoreAllMocks();
});



it("renders <Profile />", async () => {

  // Mock the queried profile.
  const mockProfile = {
    username: "username1",
    email: "email@abc.com",
    hobby: "poker"
  };

  // Mock that the authenticated user is not following the queried user profile.
  jest.spyOn(global, "fetch").mockResolvedValue(Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      msg: "success",
      profile: mockProfile
    })
  }));


  // Render with redux and the store.
  act(() => {
    renderWithProviders(
      <AlertNotificationsProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </AlertNotificationsProvider>
    );
  });


  // Initially, the spinner should be rendered.
  expect(screen.getByTestId("theSpinner")).toBeInTheDocument();

  expect(fetch).toHaveBeenCalledTimes(1);

  // After the backend request, the username and hobby elements should be rendered.
  const usernameElement = await screen.findByText(`@${mockProfile.username}`);
  expect(usernameElement).toBeInTheDocument();

  const hobbyElement = await screen.findByText(mockProfile.hobby);
  expect(hobbyElement).toBeInTheDocument();

});


it("renders <Profile /> container correctly", async () => {

  // Mock the queried profile.
  const mockProfile = {
    username: "username1",
    email: "email@abc.com",
    hobby: "poker"
  };

  // Mock that the authenticated user is not following the queried user profile.
  jest.spyOn(global, "fetch").mockResolvedValue(Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      msg: "success",
      profile: mockProfile
    })
  }));


  // Render with redux and the store.
  let container;
  act(() => {
    const { asFragment } = renderWithProviders(
      <AlertNotificationsProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </AlertNotificationsProvider>
    );
    container = asFragment;
  });


  expect(fetch).toHaveBeenCalledTimes(1);

  // After the backend request, the username should be rendered.
  const usernameElement = await screen.findByText(`@${mockProfile.username}`);
  expect(usernameElement).toBeInTheDocument();


  // Expect that the rendered container matches the snapshot.
  expect(container()).toMatchSnapshot();

});


it("renders the follow button if the authenticated user is not already following the queried user profile", async () => {

  // Mock the queried profile.
  const mockProfile = {
    username: "username1",
    email: "email@abc.com",
    hobby: "poker"
  };

  // Mock that the authenticated user is not following the queried user profile.
  const isAuthFollowingTheUser = false;

  // Mock the querying of a user profile.
  jest.spyOn(global, "fetch").mockResolvedValue(Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      msg: "success",
      profile: mockProfile,
      isAuthFollowingTheUser
    })
  }));


  // Render the Profile component with redux, store, and other providers.
  act(() => {
    renderWithProviders(
      <AlertNotificationsProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </AlertNotificationsProvider>
    );
  });


  // Reference the follow button.
  const followButton = await screen.findByText("Follow");

  // Assert
  expect(followButton).toBeInTheDocument();

});


it("renders the unfollow button if the authenticated user is following the queried user profile", async () => {

  // Mock the queried profile.
  const mockProfile = {
    username: "username1",
    email: "email@abc.com",
    hobby: "poker"
  };

  // Mock that the authenticated user is following the queried user profile.
  const isAuthFollowingTheUser = true;

  // Mock the querying of a user profile.
  jest.spyOn(global, "fetch").mockResolvedValue(Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      msg: "success",
      profile: mockProfile,
      isAuthFollowingTheUser
    })
  }));


  // Render the Profile component with redux, store, and other providers.
  act(() => {
    renderWithProviders(
      <AlertNotificationsProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </AlertNotificationsProvider>
    );
  });


  // Reference the Unfollow button.
  const unfollowButton = await screen.findByText("Unfollow");

  // Assert
  expect(unfollowButton).toBeInTheDocument();

});


it("replaces the follow button with unfollow button once the authenticated user successfully follows another user", async () => {

  // Mock the queried profile.
  const mockProfile = {
    username: "username1",
    email: "email@abc.com",
    hobby: "poker"
  };

  // Mock that the authenticated user is initially not following the queried user profile.
  const isAuthFollowingTheUser = false;

  // Mock the querying of a user profile.
  jest.spyOn(global, "fetch").mockResolvedValueOnce(Promise.resolve({
    ok: true,
    status: 200,
    json: async () => ({
      msg: "success",
      profile: mockProfile,
      isAuthFollowingTheUser
    })
  }));


  // Mock the fetch request of the authenticated user following the queried user profile.
  jest.spyOn(global, "fetch").mockResolvedValueOnce(Promise.resolve({
    status: 201,
    json: async () => ({
      msg: "success"
    })
  }));


  // Render the Profile component with redux, store, and other providers.
  act(() => {
    renderWithProviders(
      <AlertNotificationsProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </AlertNotificationsProvider>
    );
  });


  // On initial render of the page, reference the follow button.
  const followButton = await screen.findByText("Follow");

  // Assert
  expect(followButton).toBeInTheDocument();

  // Simulate the click event on the follow button.
  fireEvent.click(followButton);

  // Wait for the follow request to complete and then assert that the unfollow button is rendered.
  const unfollowButton = await screen.findByText("Unfollow");

  // Assert
  expect(unfollowButton).toBeInTheDocument();


});