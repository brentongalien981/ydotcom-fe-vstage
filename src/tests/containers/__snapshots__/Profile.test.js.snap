// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Profile /> container correctly 1`] = `
<DocumentFragment>
  <div
    class="profile-container"
  >
    <h4
      class="page-title"
    >
      Profile
    </h4>
    <div
      class="profile-sections"
    >
      <img
        class="profile-photo rounded-circle"
        src="/photos/penguin2.jpg"
      />
    </div>
    <div
      class="profile-sections"
    >
      @username1
    </div>
    <div
      class="profile-sections"
    >
      <button
        class="btn btn-primary btn btn-primary"
        type="button"
      >
        Follow
      </button>
    </div>
    <ul
      class="mb-3 profile-sections nav nav-tabs nav-fill"
      role="tablist"
    >
      <li
        class="nav-item"
        role="presentation"
      >
        <button
          aria-controls="react-aria-:r3:-tabpane-bio"
          aria-selected="true"
          class="nav-link active"
          data-rr-ui-event-key="bio"
          id="react-aria-:r3:-tab-bio"
          role="tab"
          type="button"
        >
          Bio
        </button>
      </li>
      <li
        class="nav-item"
        role="presentation"
      >
        <button
          aria-controls="react-aria-:r3:-tabpane-posts"
          aria-selected="false"
          class="nav-link"
          data-rr-ui-event-key="posts"
          id="react-aria-:r3:-tab-posts"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Posts
        </button>
      </li>
      <li
        class="nav-item"
        role="presentation"
      >
        <button
          aria-controls="react-aria-:r3:-tabpane-followers"
          aria-selected="false"
          class="nav-link"
          data-rr-ui-event-key="followers"
          id="react-aria-:r3:-tab-followers"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Followers
        </button>
      </li>
      <li
        class="nav-item"
        role="presentation"
      >
        <button
          aria-controls="react-aria-:r3:-tabpane-following"
          aria-selected="false"
          class="nav-link"
          data-rr-ui-event-key="following"
          id="react-aria-:r3:-tab-following"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Following
        </button>
      </li>
    </ul>
    <div
      class="tab-content"
    >
      <div
        aria-labelledby="react-aria-:r3:-tab-bio"
        class="fade tab-pane active show"
        id="react-aria-:r3:-tabpane-bio"
        role="tabpanel"
      >
        <div
          class="d-flex justify-content-center profile-bio"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  profession:
                </td>
                <td />
              </tr>
              <tr>
                <td>
                  email:
                </td>
                <td>
                  email@abc.com
                </td>
              </tr>
              <tr>
                <td>
                  country:
                </td>
                <td />
              </tr>
              <tr>
                <td>
                  hobby:
                </td>
                <td>
                  poker
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        aria-labelledby="react-aria-:r3:-tab-posts"
        class="fade tab-pane"
        id="react-aria-:r3:-tabpane-posts"
        role="tabpanel"
      >
        Tab content for Posts
      </div>
      <div
        aria-labelledby="react-aria-:r3:-tab-followers"
        class="fade tab-pane"
        id="react-aria-:r3:-tabpane-followers"
        role="tabpanel"
      >
        Tab content for Followers
      </div>
      <div
        aria-labelledby="react-aria-:r3:-tab-following"
        class="fade tab-pane"
        id="react-aria-:r3:-tabpane-following"
        role="tabpanel"
      >
        Tab content for Following
      </div>
    </div>
  </div>
</DocumentFragment>
`;
